src/SonarQube.Client/Services/SonarQubeService.cs

```diff
@@ -0,0 +1,231 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarQube.Client.Messages;
+using SonarQube.Client.Models;
+
+namespace SonarQube.Client.Services
+{
+    public class SonarQubeService : ISonarQubeService
+    {
+        internal const int MaximumPageSize = 500;
+        internal readonly Version OrganizationsFeatureMinimalVersion = new Version(6, 2);
+
+        private readonly ISonarQubeClient sonarqubeClient;
+
+        private ConnectionRequest connection;
+        private bool isConnected;
+        private Version serverVersion;
+
+        public SonarQubeService(ISonarQubeClient sonarqubeClient)
+        {
+            if (sonarqubeClient == null)
+            {
+                throw new ArgumentNullException(nameof(sonarqubeClient));
+            }
+
+            this.sonarqubeClient = sonarqubeClient;
+        }
+
+        public bool HasOrganizationsFeature
+        {
+            get
+            {
+                EnsureIsConnected();
+
+                return this.serverVersion >= OrganizationsFeatureMinimalVersion;
+            }
+        }
+
+        public async Task ConnectAsync(ConnectionInformation connection, CancellationToken token)
+        {
+            if (this.isConnected)
+            {
+                throw new InvalidOperationException("This operation expects the service not to be connected.");
+            }
+
+            var ConnectionRequest = new ConnectionRequest
+            {
+                Authentication = connection.Authentication,
+                ServerUri = connection.ServerUri,
+                Login = connection.UserName,
+                Password = connection.Password
+            };
+
+            var credentialsValidationResult = await this.sonarqubeClient.ValidateCredentialsAsync(ConnectionRequest, token);
+            credentialsValidationResult.EnsureSuccess();
+
+            if (!credentialsValidationResult.Value.IsValid)
+            {
+                throw new Exception("Invalid credentials."); // TODO: Provide better exception
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') 'System.Exception' should not be thrown by user code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS112)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)